app/models/game.rb

```diff
@@ -49,9 +49,62 @@ def render_piece(x, y)
 
   def check?(is_white)
     king = pieces.find_by(type: KING, is_white: is_white)
+    return false if king.nil?
     pieces.where(game_id: id, is_white: !is_white).find_each do |piece|
+      next if piece.x_position.nil? || piece.y_position.nil?
       return true if piece.valid_move?(king.x_position, king.y_position)
     end
     false
   end
+
+  def stalemate?(is_white)
+    return false if check?(is_white)
+    (1..8).each do |new_x|
+      (1..8).each do |new_y|
+        pieces.where(is_white: is_white).find_each do |piece|
+          return false if legal_move?(piece, new_x, new_y)
+        end
+      end
+    end
+    true
+  end
+
+  def legal_move?(piece, new_x, new_y)
+    return false unless piece.actual_move?(new_x, new_y)
+    return_val = false
+    piece_moved_start_x = piece.x_position
+    piece_moved_start_y = piece.y_position
+    piece_captured = nil
+    piece_captured_x = nil
+    piece_captured_y = nil
+    # check if you are moving pawn in en passant capture of enemy pawn
+    if piece.type == PAWN && !piece.square_occupied?(new_x, new_y)
+      if (new_x - piece_moved_start_x).abs == 1 && (new_y - piece_moved_start_y).abs == 1
+        piece_captured = Piece.get_piece_at_coor(new_x, piece_moved_start_y)
+        piece_captured_x = new_x
+        piece_captured_y = piece_moved_start_y
+      end
+    end
+    # return false if move is invalid for this piece for any of the reasons checked in piece #valid_move?
+    return false unless piece.valid_move?(new_x, new_y)
+    # If square is occupied, respond according to whether piece is occupied by friend or foe
+    if piece.square_occupied?(new_x, new_y)
+      occupying_piece = Piece.get_piece_at_coor(new_x, new_y)
```

Comment:
Does this only get pieces from the current game? Looks like it looks at pieces that belong to any game